lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -3621,23 +3630,42 @@ static bool ReplaceUseOfZeroInitPostDom(Instruction *I, Value *V,
   }
   return true;
 }
+
+// Recursively collect all successors of BB and BB's successors.
+// BB will not be in set unless it's reachable through its successors.
+static void CollectReachableBBs(BasicBlock *BB, SmallPtrSet<BasicBlock*, 8> &Reachable) {
+  for (auto S : successors(BB)) {
+    if (Reachable.insert(S).second)
+      CollectReachableBBs(S, Reachable);
+  }
+}
+
 // When zero initialized GV has only one define, all uses before the def should
 // use zero.
 static bool ReplaceUseOfZeroInitBeforeDef(Instruction *I, GlobalVariable *GV) {
   BasicBlock *BB = I->getParent();
   Function *F = I->getParent()->getParent();
   // Make sure I is the last inst for BB.
+  BasicBlock *NewBB = nullptr;
   if (I != BB->getTerminator())
-    BB->splitBasicBlock(I->getNextNode());
+    NewBB = BB->splitBasicBlock(I->getNextNode());
 
+  bool bSuccess = false;
   if (&F->getEntryBlock() == I->getParent()) {
-    return ReplaceUseOfZeroInitEntry(I, GV);
+    bSuccess = ReplaceUseOfZeroInitEntry(I, GV);
   } else {
-    // Post dominator tree.
-    PostDominatorTree PDT;
-    PDT.runOnFunction(*F);
-    return ReplaceUseOfZeroInitPostDom(I, GV, PDT);
+    DominatorTree DT;
+    DT.recalculate(*F);
```

Comment:
We ran into compile time performance issues recalculating this redundantly elsewhere. Since this is called from LowerMemCpy, which already has a DT for this function, you might as well pass it in and reuse it.

Reply:
Unfortunately, it will only work if updated for the split block, then again for the re-merge.  That is possible, but I will need to be cautious since I haven't written code that updates the Dom this way.

The other possibility is get rid of the block split operation and do instruction iteration inside the block with the memcpy to determine before/after status for instructions in that block.  I think I would prefer instruction iteration if we are going to try optimizing this, since splitting and joining blocks will have to iterate the instructions in the block in any case (at least the ones following the memcpy).  Additionally, the full dom tree is not necessary, nor is full reachable list.  These could be built manually and cached on the way from the blocks containing users by following predecessors.  One then just keeps track of reachable and dominated properties for each cached block.  Even if we use the existing dom tree, we still need reachability, which could have been computed in parallel.

Perhaps we can leave optimization of this to another step (this was already computing post-dom tree on the fly before), and then there are other higher-level refactor changes that might make the local exercise moot anyway.

Reply:
Fair enough. If it was easy I wanted to suggest it. It's clearly not and, as you say, the price was being paid before so we're no worse off